File: Leetcode July Challenge/Python/4 Sum.py
result=[]
        nums.sort()
        
        for i in range(len(nums)-3):
            if i>0 and nums[i]==nums[i-1]:
                continue
                
            for j in range(i+1,len(nums)-2):
                
                if j>i+1 and nums[j] == nums[j-1]:
                    continue
                sum=target-nums[i]-nums[j]
                
                l=j+1
                r=len(nums)-1
                while l<r:
                    if nums[l]+nums[r]==sum:
                        result.append([nums[i],nums[j],nums[l],nums[r]])
                        while l<r and nums[l]==nums[l+1]:
                            l+=1
                        while l<r and nums[r]==nums[r-1]:
                            r-=1
                        l+=1
                        r-=1
                    elif nums[l]+nums[r]<sum:
                        l+=1
                    else:
                        r-=1
                while j < len(nums)-3 and nums[j] == nums[j+1]:
                    j+=1
                j+=1       
        return result

File: Leetcode July Challenge/C++/4 Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int n = nums.size();
        for(int i = 0; i< n - 3; i++){
            if(i > 0 && nums[i] == nums[i - 1]){
                continue;
            }
            for(int j = i + 1; j < n-2; j++){
                if(j > i + 1 && nums[j] == nums[j - 1]){
                    continue;
                }
                
                int tSum = target - (nums[i] + nums[j]);
                int l = j + 1, h = n - 1;
                while (l < h){
                    if (nums[l] + nums[h] == tSum){
                        vector<int> oneAns = {nums[i], nums[j], nums[l], nums[h]};
                        ans.push_back(oneAns);
                        while(l < h && nums[l] == nums[l + 1]){
                            l++;
                        }
                        while(l < h && nums[h] == nums[h - 1]){
                            h--;
                        }
                        l++;
                        h--;
                    }
                    else if(nums[l] + nums[h] < tSum){
                        l++;
                    }
                    else{
                        h--;
                    }
                    
                }
                
            }
        }
          
        return ans;
    }
};
